src/decode.rs

```diff
@@ -731,32 +732,52 @@ unsafe fn read_pal_indices(
         // top/left-to-bottom/right diagonals ("wave-front")
         let first = cmp::min(i, w4 * 4 - 1);
         let last = (i + 1).checked_sub(h4 * 4).unwrap_or(0);
-        order_palette(pal_idx, stride, i, first, last, order, ctx);
+        order_palette(pal_tmp, stride, i, first, last, order, ctx);
         for (m, j) in (last..=first).rev().enumerate() {
             let color_idx = rav1d_msac_decode_symbol_adapt8(
                 &mut ts.msac,
                 &mut color_map_cdf[ctx[m] as usize],
                 pal_sz - 1,
             ) as usize;
-            pal_idx[(i - j) * stride + j] = order[m][color_idx];
+            pal_tmp[(i - j) * stride + j] = order[m][color_idx];
         }
     }
-    // fill invisible edges
+    // fill invisible edges and pack to 4-bit (2 pixels per byte)
     if bw4 > w4 {
         for y in 0..4 * h4 {
             let offset = y * stride + (4 * w4);
             let len = 4 * (bw4 - w4);
-            let filler = pal_idx[offset - 1];
-            pal_idx[offset..][..len].fill(filler);
+            let filler = pal_tmp[offset - 1];
+            pal_tmp[offset..][..len].fill(filler);
         }
     }
-    if h4 < bh4 {
-        let y_start = h4 * 4;
-        let len = bw4 * 4;
-        let (src, dests) = pal_idx.split_at_mut(stride * y_start);
-        let src = &src[stride * (y_start - 1)..][..len];
-        for y in 0..(bh4 - h4) * 4 {
-            dests[y * stride..][..len].copy_from_slice(src);
+    if let Some(pal_idx) = pal_idx {
```

Comment:
Could we just do the following and not duplicate the code?

```rust
let pal_idx = pal_idx.unwrap_or(pal_tmp);
...
```

Reply:
I had tried that, but ran into borrowing issues in `pal_idx[i] = pal_tmp[2 * i + 0] | (pal_tmp[2 * i + 1] << 4)` and ended up duplicating the code.

Reply:
Actually, the duplicated code will be put in a separate function (with asm versions) in a future PR, so this is a temporary issue

Reply:
Sounds good, I see.